Tests/Algorithm/Framework/Portfolio/MinimumVariancePortfolioOptimizerTests.cs

```diff
@@ -0,0 +1,197 @@
+            };
+        }
+
+        [TestCase(0)]
+        [TestCase(1)]
+        [TestCase(2)]
+        [TestCase(3)]
+        public void TestOptimizeWeightings(int testCaseNumber)
+        {
+            var testOptimizer = new MinimumVariancePortfolioOptimizer();
+
+            var result = testOptimizer.Optimize(
+                _historicalReturns[testCaseNumber],
+                _expectedReturns[testCaseNumber],
+                _covariances[testCaseNumber]);
+
+            Assert.AreEqual(_expectedResults[testCaseNumber], result.Select(x => Math.Round(x, 6)));
+            Assert.AreEqual(1d, result.Select(x => Math.Round(Math.Abs(x), 6)).Sum());
+        }
+
+        [TestCase(4)]
+        [TestCase(5)]
+        [TestCase(6)]
+        [TestCase(7)]
+        public void TestOptimizeWeightingsSpecifyingTargetReturns(int testCaseNumber)
+        {
+            var testOptimizer = new MinimumVariancePortfolioOptimizer(targetReturn: _targetReturns[testCaseNumber]);
+
+            var result = testOptimizer.Optimize(
+                _historicalReturns[testCaseNumber],
+                _expectedReturns[testCaseNumber],
+                _covariances[testCaseNumber]);
+
+            Assert.AreEqual(_expectedResults[testCaseNumber], result.Select(x => Math.Round(x, 6)));
+            Assert.AreEqual(1d, result.Select(x => Math.Round(Math.Abs(x), 6)).Sum());
+        }
+
+        [TestCase(0)]
+        public void EqualWeightingsWhenNoSolutionFound(int testCaseNumber)
+        {
+            var testOptimizer = new MinimumVariancePortfolioOptimizer(upper: -1);
+            var expectedResult = new double[] { 0.25, 0.25, 0.25, 0.25 };
+
+            var result = testOptimizer.Optimize(_historicalReturns[testCaseNumber]);
+
+            Assert.AreEqual(expectedResult, result);
+        }
+
+        [TestCase(0)]
+        [TestCase(1)]
+        [TestCase(2)]
+        [TestCase(3)]
+        public void BoundariesAreNotViolated(int testCaseNumber)
+        {
+            var lower = 0d;
+            var upper = 0.5d;
+            var testOptimizer = new MinimumVariancePortfolioOptimizer(lower, upper);
+
+            var result = testOptimizer.Optimize(
+                _historicalReturns[testCaseNumber],
+                _expectedReturns[testCaseNumber],
+                _covariances[testCaseNumber]);
+
+            foreach (var x in result)
+            {
+                var rounded = Math.Round(x, 6);
+                Assert.GreaterOrEqual(rounded, lower);
+                Assert.LessOrEqual(rounded, upper);
+            };
+        }
+
+        public void EmptyPortfolioReturnsEmptyArrayOfDouble()
+        {
+            var testOptimizer = new MinimumVariancePortfolioOptimizer();
+            var historicalReturns = new double[,] { { } };
+
+            var result = testOptimizer.Optimize(historicalReturns);
+
+            Assert.AreEqual(Array.Empty<double>(), result);
+        }
+
+        public void SingleSecurityPortfolioReturnsOne()
```

Comment:
These are missing the `[Test]` attribute, same for `UnconstrainedMeanVariancePortfolioOptimizer`